### PR TITLE
Enable override linker flags

### PR DESCRIPTION
**Pull request by abergmeier**
Not everyone wants to link with stdc++. Enable overriding standard link flags with LINK_FLAGS.

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by abergmeier** [comment]
@damienmg Please validate whether change will be upstreamed. Will add tests then.

**Comment by damienmg** [comment]
C++ -> Marcel :)

File: tools/cpp/unix_cc_configure.bzl [hunk 1]
```diff
@@ -84,6 +84,21 @@ def _escaped_cplus_include_paths(repository_ctx):
...
+    return [
+      "-lstdc++",
+      "-lm",  # Some systems expect -lm in addition to -lstdc++
+    ]
+
+  result = []
+  # FIXME: For paths : is already narrow but for flags it is 
+  # probably too narrow. Needs to replaced by way of properly
+  # splitting including escaping.
+  for p in repository_ctx.os.environ["LINK_FLAGS"].split(":"):
```
Comment (mhlopko) [inline tools/cpp/unix_cc_configure.bzl]: Why did yo choose ':' for separator? Is that what cmake uses? Which flags wouldn't work with naive split-on-space approach? Linker flags can be passed down as `-Wl,-foo,bar,baz` for `-foo bar baz`, right?

Also what is your reasoning behind naming these variables? Why not BAZEL_LINKER_FLAGS for consistency with your other pr?

And finally, could you add this env var to cc_configure.bzl? Thanks!

Comment (abergmeier) [reply tools/cpp/unix_cc_configure.bzl]: > Why did yo choose ':' for separator?

No particular reason. It is a bad choice though!

> Is that what cmake uses?

CMAKE uses `;`. Which is a bit more sane but still bad.

> Linker flags can be passed down as -Wl,-foo,bar,baz for -foo bar baz, right?
It _should_ work AFAIK. Seems very hackish to me. Someone with more experience might know - summoning @meteorcloudy

> Which flags wouldn't work with naive split-on-space approach?
Stuff like `-l :foo`. I vagely remember that there are some special use cases like this in gcc/ld.

> Why not BAZEL_LINKER_FLAGS for consistency with your other pr?
Well from the current set of environment variable names it is not clear to me which model they follow. +1 for `BAZEL_LINKER_FLAGS`, though.

Comment (mhlopko) [reply tools/cpp/unix_cc_configure.bzl]: I don't think `-Wl,-foo,bar,baz` is hackish, I see it used very often, internally and externally. I wouldn't worry about it. You still can express versioned libraries with `-Wl,-l,:foo` if you need to.

Yeah there's not much consistency there. BAZEL_* used to mean it's needed for bootstrapping bazel, but now it's also used to tweak windows toolchain. 

Anyway, bonus points for adding comment to cc_configure.bzl for what this variable does :)

Comment (meteorcloudy) [reply tools/cpp/unix_cc_configure.bzl]: > Linker flags can be passed down as -Wl,-foo,bar,baz for -foo bar baz, right?

The `-Wl,xxx` option for gcc passes a comma-separated list of tokens as a space-separated list of arguments to the linker. So
`gcc -Wl,aaa,bbb,ccc`
eventually becomes a linker call
`ld aaa bbb ccc`

So if this "linker_flag"s are passed to gcc, we should add `-Wl`

Comment (abergmeier) [reply tools/cpp/unix_cc_configure.bzl]: @meteorcloudy There is an awful lot of `gcc` in these sentences. I am using clang currently (which AFAIK supports a lot of strange gcc behavior) and others even stranger compilers. Are we sure the `-Wl` "variant" works for allCompilersThereAre?

Comment (meteorcloudy) [reply tools/cpp/unix_cc_configure.bzl]: IIRC, like Marcel said, we are often using this syntax to pass linker flags, so I believe it should be no problem. Marcel might know more about clang.

Comment (mhlopko) [reply tools/cpp/unix_cc_configure.bzl]: Clang is the same, that's the contract there is for passing flags from compiler driver down to the linker. I have no idea what msvc does, but also msvc toolchain autoconfiguration is not as tweakable as linux/mac toolchain, so I wouldn't worry too much about it.

**Comment by abergmeier** [comment]
Any pointers to how to so unit testing for Skylark?

**Comment by mhlopko** [comment]
No unit testing at this point :(

**Comment by mhlopko** [comment]
Ready to be reviewed again? (Feel free to ping me with PTAL when ready :)

**Comment by abergmeier** [comment]
@mhlopko Currently stuck with writing tests. For some reason `getConfiguredTarget` will not throw an exception when Skylark calls `fail`. How is one supposed to handle this properly then?

**Comment by abergmeier** [comment]
@mhlopko Depends on https://github.com/bazelbuild/bazel/pull/4180. Need to wait till the former landed.

**Comment by ulfjack** [comment]
What's up with the CLA?

**Comment by abergmeier** [comment]
> What's up with the CLA?

Surely @damienmg knows why this policy is in place.

**Comment by philwo** [comment]
@abergmeier I think the CLA bot is just alerting us that you (abergmeier) sent us a pull request that also includes commit from a different author (abergmeier-dsfishlabs). It's up to us to verify that you two are fine with getting these commits merged.

As you are the same person, I think we can be reasonably sure that you're fine with this and we can ignore the CLA bot here. ;)

**Comment by mhlopko** [comment]
And I completely forgot about this PR and implemented it myself (https://github.com/bazelbuild/bazel/commit/db80298391bd953b1a1f8e9bb83c2d4c2f3747b4). Sorry for that. I'll also implement copts side of this PR soon, so, closing. Thanks and again I'm sorry for forgetting about this!

**Pull request closed without merging**